File: packages/stunner-editors/kie-wb-common-forms/kie-wb-common-forms-core/kie-wb-common-forms-adf/kie-wb-common-forms-adf-engine/kie-wb-common-forms-adf-engine-api/src/test/java/org/kie/workbench/common/forms/adf/engine/shared/formGeneration/FormGeneratorTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *  http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License. 
 */


package org.kie.workbench.common.forms.adf.engine.shared.formGeneration;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.kie.workbench.common.forms.adf.engine.shared.FormElementFilter;
import org.kie.workbench.common.forms.adf.engine.shared.formGeneration.model.Person;
import org.kie.workbench.common.forms.adf.engine.shared.test.AbstractFormGenerationTest;
import org.kie.workbench.common.forms.model.FormDefinition;
import org.mockito.junit.MockitoJUnitRunner;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;

@RunWith(MockitoJUnitRunner.class)
public class FormGeneratorTest extends AbstractFormGenerationTest {

    @Before
    @Override
    public void init() {
        super.init();
    }

    @Test
    public void testGenerateFormForModel() {
        FormDefinition form = generator.generateFormForModel(model);

        testGeneratedForm(form,
                          Person.class.getName());
    }

    @Test
    public void testGenerateFormForClass() {
        FormDefinition form = generator.generateFormForClass(Person.class);

        testGeneratedForm(form,
                          Person.class.getName());
    }

    @Test
    public void testGenerateFormForClassName() {
        FormDefinition form = generator.generateFormForClassName(Person.class.getName());

        testGeneratedForm(form,
                          Person.class.getName());
    }

    @Test
    public void testGenerateFormForModelWithFilters() {
        FormDefinition form = generator.generateFormForModel(model, getFilters());

        checkFormGeneratedWithFilters(form);
    }

    @Test
    public void testGenerateFormForClassWithFilters() {
        FormDefinition form = generator.generateFormForClass(Person.class, getFilters());

        checkFormGeneratedWithFilters(form);
    }

    @Test
    public void testGenerateFormForClassNameWithFilters() {
        FormDefinition form = generator.generateFormForClassName(Person.class.getName(), getFilters());

        checkFormGeneratedWithFilters(form);
    }

    protected FormElementFilter[] getFilters() {
        FormElementFilter nameFilter = new FormElementFilter("name", o -> false);
        FormElementFilter lastNameFilter = new FormElementFilter("lastName", o -> false);

        return new FormElementFilter[]{nameFilter, lastNameFilter};
    }

    protected void checkFormGeneratedWithFilters(FormDefinition formDefinition) {
        assertNotNull(formDefinition);

        assertEquals(formDefinition.getFields().size(), formDefinition.getLayoutTemplate().getRows().size());

        assertNull(formDefinition.getFieldByBinding("name"));
        assertNull(formDefinition.getFieldByBinding("lastName"));
    }
}

File: packages/form-code-generator/tests/FormGenerator.test.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { FormCodeGenerator } from "../dist/FormCodeGenerator";
import { FormAsset, FormSchema } from "../dist/types";
import { ApplyForVisaSchema, ConfirmTravelSchema, dummyPatternflyTheme } from "./__mocks__/partternfly";

describe("FormCodeGenerator tests", () => {
  describe("getFormGenerator tests", () => {
    it("Lookup existing formGenerator - patternfly", () => {
      const jbpmFormGenerator = new FormCodeGenerator(dummyPatternflyTheme);
      const formGenerator = jbpmFormGenerator.getFormGenerator("patternfly");

      expect(formGenerator).not.toBeUndefined();
      expect(formGenerator.theme).toStrictEqual("patternfly");
    });

    it("Lookup wrong formGenerator", () => {
      const jbpmFormGenerator = new FormCodeGenerator();
      expect(() => jbpmFormGenerator.getFormGenerator("wrong formGenerator type")).toThrow(
        `Unsupported form generation type: "wrong formGenerator type"`
      );
    });

    it("Register formGenerator & lookup", () => {
      const myCoolFormGenerator = {
        theme: "cool new formGenerator",
        generate: jest.fn(),
      };

      const jbpmFormGenerator = new FormCodeGenerator(dummyPatternflyTheme, myCoolFormGenerator);

      const customFormGenerator = jbpmFormGenerator.getFormGenerator("cool new formGenerator");
      expect(customFormGenerator).not.toBeUndefined();
      expect(customFormGenerator).toStrictEqual(myCoolFormGenerator);

      const patternflyFormGenerator = jbpmFormGenerator.getFormGenerator("patternfly");
      expect(patternflyFormGenerator).not.toBeUndefined();
      expect(patternflyFormGenerator).toStrictEqual(dummyPatternflyTheme);
    });
  });

  describe("generateForms tests", () => {
    it("Generate forms with wrong formGenerator type", () => {
      const jbpmFormGenerator = new FormCodeGenerator();
      expect(() =>
        jbpmFormGenerator.generateForms({
          formSchemas: [{ name: "", schema: {} }],
          theme: "wrong type",
        })
      ).toThrow('Unsupported form generation type: "wrong type"');
    });

    it("Generate forms for empty schema", () => {
      const jbpmFormGenerator = new FormCodeGenerator(dummyPatternflyTheme);
      const formAssets = jbpmFormGenerator.generateForms({
        formSchemas: [{ name: "test", schema: {} }],
        theme: "patternfly",
      });

      expect(formAssets[0]).toEqual(
        expect.objectContaining({
          formAssets: expect.objectContaining({
            id: "test",
            assetName: "test.tsx",
            config: { resources: { scripts: {}, styles: {} }, schema: "{}" },
            type: "tsx",
          }),
          formErrors: undefined,
        })
      );
    });

    it("Generate forms project with schemas", () => {
      const jbpmFormGenerator = new FormCodeGenerator(dummyPatternflyTheme);
      const formAssets = jbpmFormGenerator.generateForms({
        formSchemas: [
          { name: "Apply#For#Visa", schema: ApplyForVisaSchema },
          { name: "ConfirmTravel", schema: ConfirmTravelSchema },
        ],
        theme: "patternfly",
      });

      expect(formAssets[0]).toEqual(
        expect.objectContaining({
          formAssets: expect.objectContaining({
            id: "Apply#For#Visa",
            assetName: "Apply#For#Visa.tsx",
            config: { resources: { scripts: {}, styles: {} }, schema: JSON.stringify(ApplyForVisaSchema) },
            type: "tsx",
          }),
          formErrors: undefined,
        })
      );
      expect(formAssets[1]).toEqual(
        expect.objectContaining({
          formAssets: expect.objectContaining({
            id: "ConfirmTravel",
            assetName: "ConfirmTravel.tsx",
            config: { resources: { scripts: {}, styles: {} }, schema: JSON.stringify(ConfirmTravelSchema) },
            type: "tsx",
          }),
          formErrors: undefined,
        })
      );
    });

    it("Generate forms project with schemas and one failure", () => {
      const jbpmFormGenerator = new FormCodeGenerator({
        theme: "cool formGenerator",

        generate(schema: FormSchema): FormAsset<"txt"> {
          if (schema.name === "ApplyForVisa") {
            throw new Error("Unexpected Error!");
          }

          return {
            id: schema.name,
            content: schema.name,
            type: "txt",
            assetName: `${schema.name}.txt`,
            config: {
              schema: "",
              resources: { styles: {}, scripts: {} },
            },
          };
        },
      });

      const formAssets = jbpmFormGenerator.generateForms({
        formSchemas: [
          { name: "ApplyForVisa", schema: ApplyForVisaSchema },
          { name: "ConfirmTravel", schema: ConfirmTravelSchema },
        ],
        theme: "cool formGenerator",
      });

      expect(formAssets[0]).toEqual({
        formAssets: undefined,
        formErrors: new Error("Unexpected Error!"),
      });
      expect(formAssets[1]).toEqual(
        expect.objectContaining({
          formAssets: expect.objectContaining({
            id: "ConfirmTravel",
            assetName: "ConfirmTravel.txt",
            config: { resources: { scripts: {}, styles: {} }, schema: "" },
            type: "txt",
          }),
          formErrors: undefined,
        })
      );
    });
  });
});
